blog/content/post/2024-08-19-secureboot-freebsd.md

```diff
@@ -0,0 +1,397 @@
+author: filip.lewinski
+layout: post
+published: true
+date: 2024-10-14
+archives: "2024"
+
+tags:
+  - bsd
+  - dasharo
+  - edk2
+  - secure-boot
+  - protectli
+  - uefi
+categories:
+  - Firmware
+  - IoT
+  - Miscellaneous
+  - OS Dev
+  - App Dev
+  - Security
+  - Manufacturing
+
+---
+
+## Introduction
+
+## UEFI Secure Boot
+
+UEFI Secure Boot is a security feature designed to ensure that only trusted,
+signed software can execute during the boot process, protecting against malware
+and unauthorized code execution. It verifies the signatures of bootloaders and
+kernel binaries against trusted certificates embedded in the firmware.
+
+### Linux
+
+Most distributions took the approach of using a small, signed bootloader called
+`shim`, which is pre-signed by Microsoft. Shim loads the bootloader, which
+is verified against a certificate embedded in shim, and the bootloader then
+loads the Linux kernel, also ensuring it is signed and trusted. This way, they
+can sign their kernel themselves, instead of having to ask Microsoft to sign
+each kernel update.
+
+Please see [this blogpost](https://mjg59.dreamwidth.org/20303.html) for more
+information on this specific subject.
+
+### UEFI Secure Boot status on FreeBSD - brief summary
+
+In short, the [FreeBSD wiki](https://wiki.freebsd.org/SecureBoot) proposes two
+stages of Secure Boot implementation.
+
+First would be to enable booting the system with Microsoft-signed shim and a
+FreeBSD-signed EFI loader, while also offering users to generate their own
+keys and certificates.
+
+The second stage would involve having all of the drivers and kernel modules to
+be signed and authorized, resulting in a locked-down system akin to Fedora's
+implementation.
+
+As of the time of writing this post, the first stage is mostly complete. It is
+possible to create and sign a complete EFI executable containing the bootloader
+and the FreeBSD kernel, which can be booted using Dasharo firmware Secure Boot.
+
+Now, let's analyze what the said executable consists of.
+
+#### FreeBSD UEFI boot process
+
+The regular UEFI boot process on FreeBSD is as follows:
+
+* UEFI firmware loads `boot1.efi`
+* `boot1.efi` in turn loads `loader.efi`
+* Finally, `loader.efi` loads the kernel.
+
+The target scenario, according to the FreeBSD Foundation, would be that a
+cryptographic handshake would verify each transition within the process. That
+is still under development, and the currently available option is a little less
+modular.
+
+It is possible to skip `boot1.efi` and combine the loader and the kernel into
+one large package, which then can be signed with self-issued keys. The keys
+are enrolled into Dasharo firmware, which boots the loader-kernel object
+directly.
+
+## Creating a Secure Boot - ready FreeBSD EFI executable
+
+Please note that this guide assumes that you have a working FreeBSD
+installation that you wish to modify for Secure Boot support. Installation of
+the system will not be covered here.
```

Comment:
Which version of FreeBSD was tested here?

Reply:
[Specified](https://github.com/3mdeb/news-and-ideas/pull/603/commits/dc0833e717f8971a7ea1d563294813d6a1406384#diff-1e142bc7519e97361fc0d3867abcb7fa3d257a4b5ccb79795e5e00acebf7d22fR108)